async_fuse/src/fuse/decoder.rs

```diff
@@ -0,0 +1,339 @@
+
+        check_size(self.bytes.len(), ty_size)?;
+        check_align::<T>(self.bytes.as_ptr())?;
+
+        unsafe {
+            let bytes = self.pop_bytes_unchecked(ty_size);
+            Ok(&*(bytes.as_ptr().cast()))
+        }
+    }
+
+    /// Fetch a slice of target instances with the length `n`
+    #[allow(dead_code)]
+    pub fn fetch_slice<T: FuseAbiData + Sized>(
+        &mut self,
+        n: usize,
+    ) -> Result<&'b [T], DecodeError> {
+        let ty_size: usize = mem::size_of::<T>();
+        let ty_align: usize = mem::align_of::<T>();
+        debug_assert!(ty_size > 0 && ty_size.wrapping_rem(ty_align) == 0);
+
+        let (size, is_overflow) = ty_size.overflowing_mul(n);
+        if is_overflow {
+            trace!("number overflow: ty_size = {}, n = {}", ty_size, n);
+            return Err(DecodeError::NumOverflow);
+        }
+
+        check_size(self.bytes.len(), size)?;
+        check_align::<T>(self.bytes.as_ptr())?;
+
+        unsafe {
+            let bytes = self.pop_bytes_unchecked(size);
+            let base: *const T = bytes.as_ptr().cast();
+            Ok(slice::from_raw_parts(base, n))
+        }
+    }
+
+    /// Fetch remaining bytes and build a slice of target instances
+    pub fn fetch_all_as_slice<T: FuseAbiData + Sized>(&mut self) -> Result<&'b [T], DecodeError> {
+        let ty_size: usize = mem::size_of::<T>();
+        let ty_align: usize = mem::align_of::<T>();
+        debug_assert!(ty_size > 0 && ty_size.wrapping_rem(ty_align) == 0);
+
+        if self.bytes.len() < ty_size || self.bytes.len().wrapping_rem(ty_size) != 0 {
+            trace!(
+                "no enough bytes to fetch, remaining {} bytes but to fetch (n * {}) bytes",
+                self.bytes.len(),
+                ty_size,
+            );
+            return Err(DecodeError::NotEnough);
+        }
+
+        check_align::<T>(self.bytes.as_ptr())?;
+
+        let bytes = self.fetch_all();
+        unsafe {
+            let base: *const T = bytes.as_ptr().cast();
+            let len = bytes.len().wrapping_div(ty_size);
+            Ok(slice::from_raw_parts(base, len))
+        }
+    }
+
+    /// Fetch some nul-terminated bytes.
+    ///
+    /// [`std::ffi::CStr::to_bytes`](https://doc.rust-lang.org/stable/std/ffi/struct.CStr.html#method.to_bytes)
+    /// will take O(n) time in the future.
+    ///
+    /// `slice::len` is always O(1)
+    pub fn fetch_c_str(&mut self) -> Result<&'b [u8], DecodeError> {
+        let strlen: usize = memchr(0, self.bytes)
+            .ok_or_else(|| {
+                trace!("no trailing zero in bytes, cannot fetch c-string");
+                DecodeError::NotEnough
+            })?
+            .wrapping_add(1);
+        debug_assert!(strlen <= self.bytes.len());
+        unsafe { Ok(self.pop_bytes_unchecked(strlen)) }
+    }
+
+    /// Fetch some bytes as a OS-string
+    #[cfg(any(target_os = "linux", target_os = "macos"))]
```

Comment:
We can ignore the macos curretnly.

Reply:
shall we emit a compile error when compiles on macos?
```rust
// src/lib.rs
#[cfg(target_os = "macos")]
compile_error!("async-fuse does not support macos")
```